src/persist/src/operators/stream.rs

```diff
@@ -41,46 +50,215 @@ where
+        let mut seal_op = OperatorBuilder::new(operator_name.clone(), self.scope());
+
+        let mut data_input = seal_op.new_input(&self, Pipeline);
+
+        let (mut data_output, data_output_stream) = seal_op.new_output();
+        let (mut error_output, error_output_stream) = seal_op.new_output();
+
+        let mut data_buffer = Vec::new();
+        let mut input_frontier =
+            Antichain::from_elem(<G::Timestamp as timely::progress::Timestamp>::minimum());
+
+        let active_seal_operator = self.scope().index() == 0;
+
+        seal_op.build(move |mut capabilities| {
+            let mut err_cap = Some(capabilities.pop().expect("missing capability"));
+            let mut _data_cap = capabilities.pop();
+
+            move |frontiers| {
+                let mut data_output = data_output.activate();
+                let mut error_output = error_output.activate();
+
+                // pass through data
+                data_input.for_each(|time, data| {
+                    data.swap(&mut data_buffer);
+
+                    let mut session = data_output.session(&time);
+                    session.give_vec(&mut data_buffer);
+                });
+
+                // seal if the frontier advances
+                let new_input_frontier = frontiers[0].frontier();
+                let progress =
+                    !PartialOrder::less_equal(&new_input_frontier, &input_frontier.borrow());
+
+                if progress {
+                    input_frontier.clear();
+                    input_frontier.extend(new_input_frontier.into_iter().cloned());
+
+                    if let Some(data_frontier) = input_frontier.elements().first() {
+                        let err_cap = err_cap.as_mut().expect("missing capability");
+
+                        if active_seal_operator {
+                            log::trace!("Sealing {} up to {}", &operator_name, data_frontier);
+
+                            let result = write.seal(*data_frontier).recv();
+                            if let Err(e) = result {
+                                log::error!(
+                                    "Error sealing {} up to {}: {:?}",
+                                    &operator_name,
+                                    err_cap.time(),
+                                    e
+                                );
+                                let mut session = error_output.session(&err_cap);
+                                // TODO: make error retractable? Probably not...
+                                session.give((e.to_string(), *err_cap.time(), 1));
+                            }
+
+                            let compaction_frontier = Antichain::from_elem(data_frontier - 1);
+                            // TODO: we have to be careful not to compact too far, especially not
+                            // beyond the binding seal frontier
+                            match write.allow_compaction(compaction_frontier).recv() {
```

Comment:
woah is this really how allow compaction works in sources?

Reply:
It is not! In the real code the coordinator tells workers when they can compact. See here: https://github.com/MaterializeInc/materialize/blob/41b566c1704cbf5b46f1636b98a9065e3c10e513/src/dataflow/src/server.rs#L932. We have to do something similar, where workers have a mapping to write handles for which they need to allow compaction.

I had this code in there to see actual compaction going on in the mock Kafka example. I'll remove from this code and maybe add compaction as an explicit thing in the example.

Thinking more broadly, I think with persistence we wouldn't even need the coordinator to tell the workers that they can compact. The coordinator could talk to persistence directly. Follow-up from that: if/when the coordinator, workers, ingestion, and persistence are separate (possibly distributed) processes, can there be multiple writers to a persistent collection? Right now locking would prevent that, I think?

Reply:
> I'll remove from this code and maybe add compaction as an explicit thing in the example.

Yeah this sounds right to me

> Thinking more broadly, I think with persistence we wouldn't even need the coordinator to tell the workers that they can compact. The coordinator could talk to persistence directly.

This is one of the things I'm hoping gets figured out in the architecture/correctness discussions that frank is pushing on 🤞 

> Follow-up from that: if/when the coordinator, workers, ingestion, and persistence are separate (possibly distributed) processes, can there be multiple writers to a persistent collection? Right now locking would prevent that, I think?

Writes and seals, I think no. If there's a reason you think we need that, let's definitely jump on this sooner than later because that's a huge violation of assumptions I'm making.

allow_compaction is possibly a different story: IIUC the way this works in arranged/trace is that the compaction frontier is only advanced to some point once all readers have advanced it to that point. It's easy to imagine that we'd want our analog to match. I'm sort of hoping to punt on this for a while while being prepared to revisit it if in a services world if necessary. 

One complication here is the idea of arjun's chain of materialize instances (also applicable to blue/green deployments etc). A chained mz instance using a persistent collection as a source IMO should not be able to _in any way_ affect the performance of the cluster that's writing the collection. That includes holding up compactions